buildpack/files.go

```diff
@@ -192,9 +197,57 @@ func containsName(unmet []Unmet, name string) bool {
 
 // layer content metadata
 
-type LayerMetadataFile struct {
-	Data   interface{} `json:"data" toml:"metadata"`
-	Build  bool        `json:"build" toml:"build"`
-	Launch bool        `json:"launch" toml:"launch"`
-	Cache  bool        `json:"cache" toml:"cache"`
+type EncoderDecoder interface {
+	IsSupported(buildpackAPI string) bool
+	Encode(file *os.File, lmf layertypes.LayerMetadataFile) error
+	Decode(path string) (layertypes.LayerMetadataFile, string, error)
+}
+
+func EncodeFalseFlags(lmf layertypes.LayerMetadataFile, path, buildpackAPI string) error {
+	fh, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+
+	lmf.UnsetFlags()
+
+	encoders := []EncoderDecoder{
+		v05.NewEncoderDecoder(),
+		v06.NewEncoderDecoder(),
+	}
+
+	for _, encoder := range encoders {
+		if encoder.IsSupported(buildpackAPI) {
+			return encoder.Encode(fh, lmf)
+		}
+	}
+	return errors.New("couldn't find an encoder")
+}
+
+func DecodeLayerMetadataFile(path, buildpackAPI string) (layertypes.LayerMetadataFile, string /*warning/error*/, error) {
```

Comment:
This isn't something that should change in this PR, but I wonder if we should introduce a pattern for raising errors that could be sometimes emitted as logs/warnings. The caller code might look like:

```go
decodedTOML, err := decoder.Decode(path, buildpackAPI); 
if err != nil {
  if errors.IsTerminal(err) { 
    return err
  }
  out.Write([]byte(err.Error()))
}
... processing logic with decodedTOML...
```

Reply:
Is the `/*warning/error*/` intended?

Reply:
Yes. One of the return values is a string that will be shown as an error or a warning depending on the buildpack API version.
If you have a better comment, I would love to change the current one :)

Reply:
Hmm... why couldn't `Decoder.Decode` return the message in an error when it is supposed to be an error?

Reply:
Because for buildpack API < 0.6, we return a warning in this case: https://github.com/buildpacks/lifecycle/pull/537/files#diff-e348e4b0ed5d889979d2fb076558d56656b133c84892137fdd110002e8948ec8R113-R116

Reply:
I added a TODO comment in the following commit: https://github.com/buildpacks/lifecycle/pull/537/commits/3ca28cd072aa3714b4528c4cf7062f7ef0d094c4.
I thought about adding the `Logger` to the `DecodeLayerMetadataFile` function and then for `buildpack API < 0.6`, we'll use the logger to print a warning if needed (in case the flags are in a types table), and for `buildpack API >= 0.6`, we'll just return an error if needed (in case the flags are at the top level).
The only "problem" is that the `Logger` interface is defined in the `buildpack` package and we can't import it to the `v05` and `v06` packages because we'll have circular imports.
We can redefine the `Logger` interface in each of these packages, but I think that instead we should use the `layertypes` package and define it once there. We should probably rename this package and we can also decide to make it an "interfaces package".
I'll leave it for a different PR.